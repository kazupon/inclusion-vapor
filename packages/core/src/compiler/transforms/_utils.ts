import { parse } from 'svelte/compiler'
import { generate } from '@vue-vapor/compiler-vapor'
import { transform } from '../transform'
import { IRNodeTypes } from '../ir'

import type {
  RootIRNode as VaporRootIRNode,
  CompilerOptions as VaporCompilerOptions
} from '@vue-vapor/compiler-vapor'
import type { CompilerOptions } from '../compile'
import type { RootNode } from '../ir'

export const DEFAULT_OPTIONS: CompilerOptions = {
  prefixIdentifiers: true
}
export const DEFAULT_VAPOR_COMPILER_OPTIONS = DEFAULT_OPTIONS as VaporCompilerOptions

export function makeCompile(options: CompilerOptions = {}) {
  return (source: string, overrideOptions: CompilerOptions = {}) => {
    const svelteAst = parse(source)

    const ast: RootNode = {
      type: IRNodeTypes.ROOT,
      children: svelteAst.html.children || [],
      source,
      components: [],
      directives: [],
      helpers: new Set(),
      temps: 0
    }

    const ir = transform(ast, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })
    const { code, helpers, vaporHelpers } = generate(ir as unknown as VaporRootIRNode, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })

    return { ast, ir, code, helpers, vaporHelpers }
  }
}
