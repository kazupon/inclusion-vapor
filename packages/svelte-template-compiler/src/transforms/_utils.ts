import { parse } from 'svelte/compiler'
import { generate } from '@vue-vapor/compiler-vapor'
import { transform } from '../transform'
import { IRNodeTypes } from '../ir'

import type {
  RootIRNode as VaporRootIRNode,
  CompilerOptions as VaporCompilerOptions
} from '@vue-vapor/compiler-vapor'
import type { CompilerOptions } from '../compile'
import type { RootNode, RootIRNode } from '../ir'

export const DEFAULT_OPTIONS: CompilerOptions = {
  prefixIdentifiers: true
}
export const DEFAULT_VAPOR_COMPILER_OPTIONS = DEFAULT_OPTIONS as VaporCompilerOptions

export function makeCompile(options: CompilerOptions = {}) {
  return (
    source: string,
    overrideOptions: CompilerOptions = {}
  ): {
    ast: RootNode
    ir: RootIRNode
    code: string
    helpers: Set<string>
    vaporHelpers: Set<string>
  } => {
    const svelteAst = parse(source)

    const ast: RootNode = {
      type: IRNodeTypes.ROOT,
      children: svelteAst.html.children || [],
      source,
      components: [],
      directives: [],
      helpers: new Set(),
      temps: 0
    }

    const ir = transform(ast, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })
    const { code, helpers, vaporHelpers } = generate(ir as unknown as VaporRootIRNode, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })

    return { ast, ir, code, helpers, vaporHelpers }
  }
}
