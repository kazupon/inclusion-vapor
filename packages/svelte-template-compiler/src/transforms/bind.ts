// SPDX-License-Identifier: MIT
// Modifier: kazuya kawaguchi (a.k.a. kazupon)
// Forked from `@vue/compiler-vapor`
// Author: Evan you (https://github.com/yyx990803) and Vapor team (https://github.com/orgs/vuejs/teams/vapor)
// Repository url: https://github.com/vuejs/core-vapor
// Code url: https://github.com/vuejs/core-vapor/blob/6608bb31973d35973428cae4fbd62026db068365/packages/compiler-vapor/src/transforms/vBind.ts

import {
  ErrorCodes,
  NodeTypes,
  createCompilerError,
  createSimpleExpression
} from '@vue-vapor/compiler-dom'
import { camelize, extend } from '@vue-vapor/shared'
import { isReservedProp, resolveExpression } from './utils.ts'

import type { SimpleExpressionNode } from '@vue-vapor/compiler-dom'
import type { TransformContext } from './context.ts'
import type { DirectiveTransform } from './types.ts'

// same-name shorthand - :arg is expanded to :arg="arg"
export function normalizeBindShorthand(
  arg: SimpleExpressionNode,
  context: TransformContext
): SimpleExpressionNode {
  if (arg.type !== NodeTypes.SIMPLE_EXPRESSION || !arg.isStatic) {
    // only simple expression is allowed for same-name shorthand
    context.options.onError(
      createCompilerError(ErrorCodes.X_V_BIND_INVALID_SAME_NAME_ARGUMENT, arg.loc)
    )
    return createSimpleExpression('', true, arg.loc)
  }

  const propName = camelize(arg.content)
  const exp = createSimpleExpression(propName, false, arg.loc)
  exp.ast = null // eslint-disable-line unicorn/no-null
  return exp
}

export const transformVBind: DirectiveTransform = (dir, _node, context) => {
  const { loc, modifiers } = dir
  let { exp } = dir
  let arg = dir.arg!
  const modifiersString = new Set(modifiers.map(s => s.content))

  if (!exp) {
    exp = normalizeBindShorthand(arg, context)
  }
  if (!exp.content.trim()) {
    context.options.onError(createCompilerError(ErrorCodes.X_V_BIND_NO_EXPRESSION, loc))
    exp = createSimpleExpression('', true, loc)
  }

  exp = resolveExpression(exp)
  arg = resolveExpression(arg)

  if (arg.isStatic && isReservedProp(arg.content)) {
    return
  }

  let camel = false
  if (modifiersString.has('camel')) {
    if (arg.isStatic) {
      arg = extend({}, arg, { content: camelize(arg.content) })
    } else {
      camel = true
    }
  }

  return {
    key: arg,
    value: exp,
    loc,
    runtimeCamelize: camel,
    modifier: modifiersString.has('prop') ? '.' : modifiersString.has('attr') ? '^' : undefined
  }
}
