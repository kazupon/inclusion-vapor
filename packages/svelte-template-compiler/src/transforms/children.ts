// SPDX-License-Identifier: MIT
// Modifier: kazuya kawaguchi (a.k.a. kazupon)
// Forked from `@vue/compiler-vapor`
// Author: Evan you (https://github.com/yyx990803) and Vapor team (https://github.com/orgs/vuejs/teams/vapor)
// Repository url: https://github.com/vuejs/core-vapor
// Code url: https://github.com/vuejs/core-vapor/blob/6608bb31973d35973428cae4fbd62026db068365/packages/compiler-vapor/src/transforms/transformChildren.ts

import { DynamicFlag, IRNodeTypes, isSvelteText } from '../ir/index.ts'
import { transformNode } from '../transform.ts'

import type { IRDynamicInfo, SvelteElement } from '../ir/index.ts'
import type { TransformContext } from './context.ts'
import type { NodeTransform } from './types.ts'

export const transformChildren: NodeTransform = (node, context) => {
  const isFragment =
    node.type === IRNodeTypes.ROOT ||
    node.type === 'Fragment' ||
    node.type === 'InlineComponent' ||
    node.type === 'Slot'
  if (node.type !== 'Element' && !isFragment) {
    return
  }
  processChildren(node, context, isFragment)
}

export function processChildren(
  node: Parameters<NodeTransform>[0],
  context: Parameters<NodeTransform>[1],
  isFragment: boolean
): ReturnType<NodeTransform> {
  const children = node.children || []

  // normalize svelte text
  for (const [index, child] of [...children].entries()) {
    if (isSvelteText(child) && !child.data.trim()) {
      child.data = ' '
      if (!index) {
        children.splice(0, 1)
      } else if (index === children.length) {
        children.splice(-1, 1)
      }
    }
  }

  for (const [i, child] of children.entries()) {
    const childContext = context.create(child, i)
    transformNode(childContext)

    if (isFragment) {
      childContext.reference()
      childContext.registerTemplate()

      if (
        !(childContext.dynamic.flags & DynamicFlag.NON_TEMPLATE) ||
        childContext.dynamic.flags & DynamicFlag.INSERT
      ) {
        context.block.returns.push(childContext.dynamic.id!)
      }
    } else {
      context.childrenTemplate.push(childContext.template)
    }

    context.dynamic.children[i] = childContext.dynamic
  }

  if (!isFragment) {
    processDynamicChildren(context as TransformContext<SvelteElement>)
  }
}

function processDynamicChildren(context: TransformContext<SvelteElement>) {
  let prevDynamics: IRDynamicInfo[] = []
  let hasStaticTemplate = false
  const children = context.dynamic.children

  for (const [index, child] of children.entries()) {
    if (child.flags & DynamicFlag.INSERT) {
      prevDynamics.push(child)
    }

    if (!(child.flags & DynamicFlag.NON_TEMPLATE)) {
      if (prevDynamics.length > 0) {
        if (hasStaticTemplate) {
          context.childrenTemplate[index - prevDynamics.length] = `<!>`

          prevDynamics[0].flags -= DynamicFlag.NON_TEMPLATE
          const anchor = (prevDynamics[0].anchor = context.increaseId())

          context.registerOperation({
            type: IRNodeTypes.INSERT_NODE,
            elements: prevDynamics.map(child => child.id!),
            parent: context.reference(),
            anchor
          })
        } else {
          context.registerOperation({
            type: IRNodeTypes.PREPEND_NODE,
            elements: prevDynamics.map(child => child.id!),
            parent: context.reference()
          })
        }
        prevDynamics = []
      }
      hasStaticTemplate = true
    }
  }

  if (prevDynamics.length > 0) {
    context.registerOperation({
      type: IRNodeTypes.INSERT_NODE,
      elements: prevDynamics.map(child => child.id!),
      parent: context.reference()
    })
  }
}
