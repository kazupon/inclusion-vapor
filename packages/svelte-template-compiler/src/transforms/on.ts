// SPDX-License-Identifier: MIT
// Modifier: kazuya kawaguchi (a.k.a. kazupon)
// Forked from `@vue/compiler-vapor`
// Author: Evan you (https://github.com/yyx990803) and Vapor team (https://github.com/orgs/vuejs/teams/vapor)
// Repository url: https://github.com/vuejs/core-vapor
// Code url: https://github.com/vuejs/core-vapor/blob/6608bb31973d35973428cae4fbd62026db068365/packages/compiler-vapor/src/transforms/vOn.ts

import { ErrorCodes, createCompilerError, resolveModifiers } from '@vue-vapor/compiler-dom'
import { extend, makeMap } from '@vue-vapor/shared'
import { IRNodeTypes, isSvelteElement } from '../ir/index.ts'
import { EMPTY_EXPRESSION, resolveExpression } from './utils.ts'

import type { KeyOverride, SetEventIRNode } from '../ir/index.ts'
import type { DirectiveTransform } from './types.ts'

const delegatedEvents = /*#__PURE__*/ makeMap(
  'beforeinput,click,dblclick,contextmenu,focusin,focusout,input,keydown,' +
    'keyup,mousedown,mousemove,mouseout,mouseover,mouseup,pointerdown,' +
    'pointermove,pointerout,pointerover,pointerup,touchend,touchmove,' +
    'touchstart'
)

export const transformOn: DirectiveTransform = (dir, node, context) => {
  const { exp, loc, modifiers } = dir
  const isComponent = isSvelteElement(node) && node.type === 'InlineComponent'
  // TODO:
  // const isSlotOutlet = node.tag === 'slot'
  const isSlotOutlet = false

  if (!exp && modifiers.length === 0) {
    context.options.onError(createCompilerError(ErrorCodes.X_V_ON_NO_EXPRESSION, loc))
  }
  let arg = resolveExpression(dir.arg!)

  const { keyModifiers, nonKeyModifiers, eventOptionModifiers } = resolveModifiers(
    arg.isStatic ? `on${arg.content}` : arg,
    modifiers,
    null, // eslint-disable-line unicorn/no-null
    loc
  )

  let keyOverride: KeyOverride | undefined
  const isStaticClick = arg.isStatic && arg.content.toLowerCase() === 'click'
  const delegate = arg.isStatic && eventOptionModifiers.length === 0 && delegatedEvents(arg.content)

  // normalize click.right and click.middle since they don't actually fire
  if (nonKeyModifiers.includes('middle')) {
    if (keyOverride) {
      // TODO error here
    }
    if (isStaticClick) {
      arg = extend({}, arg, { content: 'mouseup' })
    } else if (!arg.isStatic) {
      keyOverride = ['click', 'mouseup']
    }
  }
  if (nonKeyModifiers.includes('right')) {
    if (isStaticClick) {
      arg = extend({}, arg, { content: 'contextmenu' })
    } else if (!arg.isStatic) {
      keyOverride = ['click', 'contextmenu']
    }
  }

  if (isComponent || isSlotOutlet) {
    const handler = exp || EMPTY_EXPRESSION
    return {
      key: arg,
      value: handler,
      handler: true
    }
  }

  const operation: SetEventIRNode = {
    type: IRNodeTypes.SET_EVENT,
    element: context.reference(),
    key: arg,
    value: exp,
    modifiers: {
      keys: keyModifiers,
      nonKeys: nonKeyModifiers,
      options: eventOptionModifiers
    },
    keyOverride,
    delegate,
    effect: !arg.isStatic
  }

  context.registerEffect([arg], operation)
}
