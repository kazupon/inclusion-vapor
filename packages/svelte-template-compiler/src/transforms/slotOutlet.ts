// SPDX-License-Identifier: MIT
// Modifier: kazuya kawaguchi (a.k.a. kazupon)
// Forked from `@vue/compiler-vapor`
// Author: Evan you (https://github.com/yyx990803) and Vapor team (https://github.com/orgs/vuejs/teams/vapor)
// Repository url: https://github.com/vuejs/core-vapor
// Code url: https://github.com/vuejs/core-vapor/blob/6608bb31973d35973428cae4fbd62026db068365/packages/compiler-vapor/src/transforms/transformSlotOutlet.ts

import { createSimpleExpression } from '@vue-vapor/compiler-dom'
import { camelize, extend } from '@vue-vapor/shared'
import {
  convertVaporDirectiveExpression,
  DynamicFlag,
  IRNodeTypes,
  isSvelteAttribute,
  isSvelteDirective,
  isSvelteElement,
  isSvelteSpreadAttribute
} from '../ir/index.ts'
import { buildProps } from './element.ts'
import { newBlock } from './utils.ts'

import type { SimpleExpressionNode } from '@vue-vapor/compiler-dom'
import type {
  BlockIRNode,
  IRProps,
  SvelteAttribute,
  SvelteBaseDirective,
  SvelteElement,
  SvelteSpreadAttribute,
  WithDirectiveIRNode
} from '../ir/index.ts'
import type { TransformContext } from './context.ts'
import type { NodeTransform } from './types.ts'

// TODO: transform vapor slot from svelte slot
// https://svelte.dev/docs/special-elements#slot
export const transformSlotOutlet: NodeTransform = (node, context) => {
  if (node.type !== 'Slot') {
    return
  }
  const { parent } = context
  if (parent && isSvelteElement(parent.node) && parent.node.type === 'InlineComponent') {
    return
  }

  const id = context.reference()
  context.dynamic.flags |= DynamicFlag.INSERT | DynamicFlag.NON_TEMPLATE
  const [fallback, exitBlock] = createFallback(
    node as SvelteElement,
    context as TransformContext<SvelteElement>
  )

  let slotName: SimpleExpressionNode | undefined
  const slotProps: (SvelteAttribute | SvelteSpreadAttribute)[] = []
  for (const attr of node.attributes as (
    | SvelteAttribute
    | SvelteSpreadAttribute
    | SvelteBaseDirective
  )[]) {
    if (isSvelteAttribute(attr)) {
      if (attr.name === 'name') {
        slotName = convertVaporDirectiveExpression(attr, { isStatic: true })
      } else {
        slotProps.push(extend({}, attr, { name: camelize(attr.name) }))
      }
    } else if (isSvelteSpreadAttribute(attr)) {
      slotProps.push(extend({}, attr))
    } else if (isSvelteDirective(attr)) {
      // TODO:
    } else {
      slotName = undefined
    }
  }

  slotName ||= createSimpleExpression('default', true)
  let irProps: IRProps[] = []
  if (slotProps.length > 0) {
    const [isDynamic, props] = buildProps(
      extend({}, node as SvelteElement, { attributes: slotProps }),
      context as TransformContext<SvelteElement>,
      true
    )
    irProps = isDynamic ? props : [props]

    const runtimeDirective = context.block.operation.find(
      (ope): ope is WithDirectiveIRNode =>
        ope.type === IRNodeTypes.WITH_DIRECTIVE && ope.element === id
    )
    if (runtimeDirective) {
      // TODO:
      // context.options.onError(
      //   createCompilerError(
      //     ErrorCodes.X_V_SLOT_UNEXPECTED_DIRECTIVE_ON_SLOT_OUTLET,
      //     runtimeDirective.dir.loc,
      //   ),
      // )
    }
  }

  return () => {
    exitBlock?.()
    context.registerOperation({
      type: IRNodeTypes.SLOT_OUTLET_NODE,
      id,
      name: slotName,
      props: irProps,
      fallback
    })
  }
}

function createFallback(
  node: SvelteElement,
  context: TransformContext<SvelteElement>
): [block?: BlockIRNode, exit?: () => void] {
  const children = node.children || []
  if (children.length === 0) {
    return []
  }

  context.node = node = extend({}, node)

  const fallback = newBlock(node)
  const exitBlock = context.enterBlock(fallback)
  context.reference()

  return [fallback, exitBlock]
}
