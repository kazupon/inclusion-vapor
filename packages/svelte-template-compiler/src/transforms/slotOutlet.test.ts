import { NodeTypes } from '@vue-vapor/compiler-dom'
import { compile as vaporCompile } from '@vue-vapor/compiler-vapor'
import { expect, test } from 'vitest'
import { IRNodeTypes } from '../ir/index.ts'
import { makeCompile } from './_utils.ts'
import { transformVBind } from './bind.ts'
import { transformChildren } from './children.ts'
import { transformComment } from './comment.ts'
import { transformElement } from './element.ts'
import { transformVOn } from './on.ts'
import { transformSlotOutlet } from './slotOutlet.ts'
import { transformText } from './text.ts'

const compileWithSlotOutlet = makeCompile({
  prefixIdentifiers: false,
  nodeTransforms: [
    transformSlotOutlet,
    transformText,
    transformElement,
    transformComment,
    transformChildren
  ],
  directiveTransforms: {
    bind: transformVBind,
    on: transformVOn
  }
})

test('default slot', () => {
  const source1 = `<slot />`
  const source2 = `<slot />`

  const { code, ir, vaporHelpers } = compileWithSlotOutlet(source1)
  const expectedResult = vaporCompile(source2)

  expect(code).toMatchSnapshot('received')
  expect(expectedResult.code).toMatchSnapshot('expected')

  expect(code).contain(`_createSlot("default", null)`)
  expect(vaporHelpers).toContain('createSlot')
  expect(ir.block.effect).toEqual([])
  expect(ir.block.operation).toMatchObject([
    {
      type: IRNodeTypes.SLOT_OUTLET_NODE,
      id: 0,
      name: {
        type: NodeTypes.SIMPLE_EXPRESSION,
        content: 'default',
        isStatic: true
      },
      props: [],
      fallback: undefined
    }
  ])
})

test('statically named slot', () => {
  const source1 = `<slot name="foo" />`
  const source2 = `<slot name="foo" />`

  const { code, ir } = compileWithSlotOutlet(source1)
  const expectedResult = vaporCompile(source2)

  expect(code).toMatchSnapshot('received')
  expect(expectedResult.code).toMatchSnapshot('expected')

  expect(code).contain(`_createSlot("foo", null)`)
  expect(ir.block.operation).toMatchObject([
    {
      type: IRNodeTypes.SLOT_OUTLET_NODE,
      id: 0,
      name: {
        type: NodeTypes.SIMPLE_EXPRESSION,
        content: 'foo',
        isStatic: true
      }
    }
  ])
})

// NOTE: svelte doesn't support dynamic slot names
test.todo('dynamically named slot', () => {
  const source1 = `<slot name={foo + bar} />`
  const source2 = `<slot :name="foo + bar" />`

  const { code, ir } = compileWithSlotOutlet(source1)
  const expectedResult = vaporCompile(source2)

  expect(code).toMatchSnapshot('received')
  expect(expectedResult.code).toMatchSnapshot('expected')

  expect(code).contain(`_createSlot(() => (foo + bar), null)`)
  expect(ir.block.operation).toMatchObject([
    {
      type: IRNodeTypes.SLOT_OUTLET_NODE,
      id: 0,
      name: {
        type: NodeTypes.SIMPLE_EXPRESSION,
        content: 'foo + bar',
        isStatic: false
      }
    }
  ])
})

test('default slot with props', () => {
  const source1 = `<slot foo="bar" baz={qux} foo-bar={fooBar} />`
  const source2 = `<slot foo="bar" :baz="qux" :foo-bar="fooBar" />`

  const { code, ir } = compileWithSlotOutlet(source1)
  const expectedResult = vaporCompile(source2)

  expect(code).toMatchSnapshot('received')
  expect(expectedResult.code).toMatchSnapshot('expected')

  expect(ir.block.operation).toMatchObject([
    {
      type: IRNodeTypes.SLOT_OUTLET_NODE,
      name: { content: 'default' },
      props: [
        [
          { key: { content: 'foo' }, values: [{ content: 'bar' }] },
          { key: { content: 'baz' }, values: [{ content: 'qux' }] },
          { key: { content: 'fooBar' }, values: [{ content: 'fooBar' }] }
        ]
      ]
    }
  ])
})

test('statically named slot with props', () => {
  const source1 = `<slot name="foo" baz={qux} foo-bar={fooBar} />`
  const source2 = `<slot name="foo" :baz="qux" :foo-bar="fooBar" />`

  const { code, ir } = compileWithSlotOutlet(source1)
  const expectedResult = vaporCompile(source2)

  expect(code).toMatchSnapshot('received')
  expect(expectedResult.code).toMatchSnapshot('expected')

  expect(ir.block.operation).toMatchObject([
    {
      type: IRNodeTypes.SLOT_OUTLET_NODE,
      name: { content: 'foo' },
      props: [
        [
          { key: { content: 'baz' }, values: [{ content: 'qux' }] },
          { key: { content: 'fooBar' }, values: [{ content: 'fooBar' }] }
        ]
      ]
    }
  ])
})

test('statically named slot with spread props', () => {
  const source1 = `<slot name="foo" {...things} />`
  const source2 = `<slot name="foo" v-bind="things" />`

  const { code, ir } = compileWithSlotOutlet(source1)
  const expectedResult = vaporCompile(source2)

  expect(code).toMatchSnapshot('received')
  expect(expectedResult.code).toMatchSnapshot('expected')

  expect(ir.block.operation).toMatchObject([
    {
      type: IRNodeTypes.SLOT_OUTLET_NODE,
      name: { content: 'foo' },
      props: [{ value: { content: 'things', isStatic: false } }]
    }
  ])
})

// NOTE: svelte doesn't support event handling on slot
test.todo('statically named slot with on:event', () => {
  const source1 = `<slot name="test" on:click={foo} on:bar={bar} baz={qux} />`
  const source2 = `<slot name="test" @click="foo" v-on="bar" :baz="qux" />`

  const { code, ir } = compileWithSlotOutlet(source1)
  const expectedResult = vaporCompile(source2)

  expect(code).toMatchSnapshot('received')
  expect(expectedResult.code).toMatchSnapshot('expected')

  expect(ir.block.operation).toMatchObject([
    {
      type: IRNodeTypes.SLOT_OUTLET_NODE,
      name: { content: 'test' },
      props: [
        [{ key: { content: 'click' }, values: [{ content: 'foo' }] }],
        { value: { content: 'bar' }, handler: true },
        [{ key: { content: 'baz' }, values: [{ content: 'qux' }] }]
      ]
    }
  ])
})

test('default slot with fallback', () => {
  const source1 = `<slot><div/></slot>`
  const source2 = `<slot><div/></slot>`

  const { code, ir } = compileWithSlotOutlet(source1)
  const expectedResult = vaporCompile(source2)

  expect(code).toMatchSnapshot('received')
  expect(expectedResult.code).toMatchSnapshot('expected')

  expect(ir.template[0]).toBe('<div></div>')
  expect(ir.block.operation).toMatchObject([
    {
      type: IRNodeTypes.SLOT_OUTLET_NODE,
      id: 0,
      name: { content: 'default' },
      fallback: {
        type: IRNodeTypes.BLOCK,
        dynamic: {
          children: [{ template: 0, id: 2 }]
        },
        returns: [2]
      }
    }
  ])
})

test('named slot with fallback', () => {
  const source1 = `<slot name="foo"><div/></slot>`
  const source2 = `<slot name="foo"><div/></slot>`

  const { code, ir } = compileWithSlotOutlet(source1)
  const expectedResult = vaporCompile(source2)

  expect(code).toMatchSnapshot('received')
  expect(expectedResult.code).toMatchSnapshot('expected')

  expect(ir.template[0]).toBe('<div></div>')
  expect(ir.block.operation).toMatchObject([
    {
      type: IRNodeTypes.SLOT_OUTLET_NODE,
      id: 0,
      name: { content: 'foo' },
      fallback: {
        type: IRNodeTypes.BLOCK,
        dynamic: {
          children: [{ template: 0, id: 2 }]
        },
        returns: [2]
      }
    }
  ])
})

test('default slot with fallback and props', () => {
  const source1 = `<slot foo={bar}><div/></slot>`
  const source2 = `<slot :foo="bar"><div/></slot>`

  const { code, ir } = compileWithSlotOutlet(source1)
  const expectedResult = vaporCompile(source2)

  expect(code).toMatchSnapshot('received')
  expect(expectedResult.code).toMatchSnapshot('expected')

  expect(ir.template[0]).toBe('<div></div>')
  expect(ir.block.operation).toMatchObject([
    {
      type: IRNodeTypes.SLOT_OUTLET_NODE,
      id: 0,
      name: { content: 'default' },
      props: [[{ key: { content: 'foo' }, values: [{ content: 'bar' }] }]],
      fallback: {
        type: IRNodeTypes.BLOCK,
        dynamic: {
          children: [{ template: 0, id: 2 }]
        },
        returns: [2]
      }
    }
  ])
})

test('named slot with fallback and props', () => {
  const source1 = `<slot name="foo" foo={bar}><div/></slot>`
  const source2 = `<slot name="foo" :foo="bar"><div/></slot>`

  const { code, ir } = compileWithSlotOutlet(source1)
  const expectedResult = vaporCompile(source2)

  expect(code).toMatchSnapshot('received')
  expect(expectedResult.code).toMatchSnapshot('expected')

  expect(ir.template[0]).toBe('<div></div>')
  expect(ir.block.operation).toMatchObject([
    {
      type: IRNodeTypes.SLOT_OUTLET_NODE,
      id: 0,
      name: { content: 'foo' },
      props: [[{ key: { content: 'foo' }, values: [{ content: 'bar' }] }]],
      fallback: {
        type: IRNodeTypes.BLOCK,
        dynamic: {
          children: [{ template: 0, id: 2 }]
        },
        returns: [2]
      }
    }
  ])
})
