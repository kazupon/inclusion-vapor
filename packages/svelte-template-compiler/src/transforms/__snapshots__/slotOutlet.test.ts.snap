// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`default slot > expected 1`] = `
"import { createSlot as _createSlot } from 'vue/vapor';

export function render(_ctx) {
  const n0 = _createSlot("default", null)
  return n0
}"
`;

exports[`default slot > received 1`] = `
"import { createSlot as _createSlot } from 'vue/vapor';

export function render(_ctx) {
  const n0 = _createSlot("default", null)
  return n0
}"
`;

exports[`default slot with fallback > expected 1`] = `
"import { createSlot as _createSlot, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")

export function render(_ctx) {
  const n0 = _createSlot("default", null, () => {
    const n2 = t0()
    return n2
  })
  return n0
}"
`;

exports[`default slot with fallback > received 1`] = `
"import { createSlot as _createSlot, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")

export function render(_ctx) {
  const n0 = _createSlot("default", null, () => {
    const n2 = t0()
    return n2
  })
  return n0
}"
`;

exports[`default slot with fallback and props > expected 1`] = `
"import { createSlot as _createSlot, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")

export function render(_ctx) {
  const n0 = _createSlot("default", [
    { foo: () => (bar) }
  ], () => {
    const n2 = t0()
    return n2
  })
  return n0
}"
`;

exports[`default slot with fallback and props > received 1`] = `
"import { createSlot as _createSlot, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")

export function render(_ctx) {
  const n0 = _createSlot("default", [
    { foo: () => (bar) }
  ], () => {
    const n2 = t0()
    return n2
  })
  return n0
}"
`;

exports[`default slot with props > expected 1`] = `
"import { createSlot as _createSlot } from 'vue/vapor';

export function render(_ctx) {
  const n0 = _createSlot("default", [
    {
      foo: () => ("bar"), 
      baz: () => (qux), 
      fooBar: () => (fooBar)
    }
  ])
  return n0
}"
`;

exports[`default slot with props > received 1`] = `
"import { createSlot as _createSlot } from 'vue/vapor';

export function render(_ctx) {
  const n0 = _createSlot("default", [
    {
      foo: () => ("bar"), 
      baz: () => (qux), 
      fooBar: () => (fooBar)
    }
  ])
  return n0
}"
`;

exports[`multiple outlets > expected 1`] = `
"import { createSlot as _createSlot, prepend as _prepend, insert as _insert, template as _template } from 'vue/vapor';
const t0 = _template("No header was provided")
const t1 = _template("<div><p>Some content between header and footer</p></div>")

export function render(_ctx) {
  const n4 = t1()
  const n0 = _createSlot("header", null, () => {
    const n2 = t0()
    return n2
  })
  const n3 = _createSlot("footer", null)
  _prepend(n4, n0)
  _insert(n3, n4)
  return n4
}"
`;

exports[`multiple outlets > received 1`] = `
"import { createSlot as _createSlot, prepend as _prepend, insert as _insert, template as _template } from 'vue/vapor';
const t0 = _template("No header was provided")
const t1 = _template("<div> <p>Some content between header and footer</p> </div>")

export function render(_ctx) {
  const n4 = t1()
  const n0 = _createSlot("header", null, () => {
    const n2 = t0()
    return n2
  })
  const n3 = _createSlot("footer", null)
  _prepend(n4, n0)
  _insert(n3, n4)
  return n4
}"
`;

exports[`named slot with fallback > expected 1`] = `
"import { createSlot as _createSlot, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")

export function render(_ctx) {
  const n0 = _createSlot("foo", null, () => {
    const n2 = t0()
    return n2
  })
  return n0
}"
`;

exports[`named slot with fallback > received 1`] = `
"import { createSlot as _createSlot, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")

export function render(_ctx) {
  const n0 = _createSlot("foo", null, () => {
    const n2 = t0()
    return n2
  })
  return n0
}"
`;

exports[`named slot with fallback and props > expected 1`] = `
"import { createSlot as _createSlot, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")

export function render(_ctx) {
  const n0 = _createSlot("foo", [
    { foo: () => (bar) }
  ], () => {
    const n2 = t0()
    return n2
  })
  return n0
}"
`;

exports[`named slot with fallback and props > received 1`] = `
"import { createSlot as _createSlot, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")

export function render(_ctx) {
  const n0 = _createSlot("foo", [
    { foo: () => (bar) }
  ], () => {
    const n2 = t0()
    return n2
  })
  return n0
}"
`;

exports[`statically named slot > expected 1`] = `
"import { createSlot as _createSlot } from 'vue/vapor';

export function render(_ctx) {
  const n0 = _createSlot("foo", null)
  return n0
}"
`;

exports[`statically named slot > received 1`] = `
"import { createSlot as _createSlot } from 'vue/vapor';

export function render(_ctx) {
  const n0 = _createSlot("foo", null)
  return n0
}"
`;

exports[`statically named slot with props > expected 1`] = `
"import { createSlot as _createSlot } from 'vue/vapor';

export function render(_ctx) {
  const n0 = _createSlot("foo", [
    {
      baz: () => (qux), 
      fooBar: () => (fooBar)
    }
  ])
  return n0
}"
`;

exports[`statically named slot with props > received 1`] = `
"import { createSlot as _createSlot } from 'vue/vapor';

export function render(_ctx) {
  const n0 = _createSlot("foo", [
    {
      baz: () => (qux), 
      fooBar: () => (fooBar)
    }
  ])
  return n0
}"
`;

exports[`statically named slot with spread props > expected 1`] = `
"import { createSlot as _createSlot } from 'vue/vapor';

export function render(_ctx) {
  const n0 = _createSlot("foo", [
    () => (things)
  ])
  return n0
}"
`;

exports[`statically named slot with spread props > received 1`] = `
"import { createSlot as _createSlot } from 'vue/vapor';

export function render(_ctx) {
  const n0 = _createSlot("foo", [
    () => (things)
  ])
  return n0
}"
`;
